File: compose_parser.py
"""Parser for Docker Compose files mapping services to LXC configurations.

This module leverages PyYAML to read Docker Compose YAML files and converts
service definitions into LXC-friendly configuration objects.  It supports
Swarm-specific extensions such as replica counts and placement constraints.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List
import logging
import os
import yaml


@dataclass
class LXCServiceConfig:
    """Simplified representation of an LXC service configuration."""
    image: str
    ports: List[str] = field(default_factory=list)
    environment: Dict[str, str] = field(default_factory=dict)
    replicas: int = 1
    constraints: List[str] = field(default_factory=list)


def _parse_environment(env: object) -> Dict[str, str]:
    """Normalise the environment section of a service definition."""
    if isinstance(env, dict):
        parsed: Dict[str, str] = {}
        for k, v in env.items():
            key = str(k)
            if v is None:
                parsed[key] = os.environ.get(key, "")
            else:
                parsed[key] = str(v)
        return parsed
    if isinstance(env, list):
        parsed: Dict[str, str] = {}
        for item in env:
            if isinstance(item, str) and "=" in item:
                key, value = item.split("=", 1)
                parsed[key] = value
            else:
                logging.warning("Ignoring malformed environment entry: %r", item)
        return parsed
    if env is not None:
        logging.warning("Unrecognised environment type: %r", type(env))
    return {}


def parse_compose(path: str) -> Dict[str, LXCServiceConfig]:
    """Parse a Docker Compose file into LXC service configurations."""
    try:
        with open(path, "r", encoding="utf-8") as fh:
            try:
                data = yaml.safe_load(fh) or {}
            except yaml.YAMLError as exc:
                raise ValueError(f"Invalid YAML in {path}: {exc}") from exc
    except OSError as exc:
        raise FileNotFoundError(f"Failed to read compose file {path}: {exc}") from exc

    services = data.get("services", {})
    configs: Dict[str, LXCServiceConfig] = {}

    for name, spec in services.items():
        image = spec.get("image", "")
        ports = spec.get("ports", [])
        environment = _parse_environment(spec.get("environment"))

        deploy = spec.get("deploy", {})
        replicas = deploy.get("replicas")
        placement = deploy.get("placement", {})
        constraints = placement.get("constraints", [])

        configs[name] = LXCServiceConfig(
            image=image,
            ports=list(ports) if isinstance(ports, list) else [],
            environment=environment,
            replicas=int(replicas) if replicas is not None else 1,
            constraints=list(constraints) if isinstance(constraints, list) else [],
        )

    return configs


__all__ = ["LXCServiceConfig", "parse_compose"]

File: src/composeParser.ts
import fs from 'fs';
import yaml from 'js-yaml';

export interface LXCServiceConfig {
  image: string;
  ports: string[];
  environment: Record<string, string>;
  replicas: number;
  constraints: string[];
}

export function parseCompose(filePath: string): Record<string, LXCServiceConfig> {
  let content: string;
  try {
    content = fs.readFileSync(filePath, 'utf8');
  } catch (err) {
    throw new Error(`Failed to read compose file at ${filePath}: ${(err as Error).message}`);
  }

  let data: any;
  try {
    data = yaml.load(content);
  } catch (err) {
    throw new Error(`Failed to parse YAML in ${filePath}: ${(err as Error).message}`);
  }

  if (typeof data !== 'object' || data === null || !('services' in data)) {
    throw new Error('Compose file missing services definition');
  }

  const services = (data as any).services as Record<string, any>;
  const result: Record<string, LXCServiceConfig> = {};

  for (const [name, svc] of Object.entries(services)) {
    const image: string = svc.image;
    const ports: string[] = Array.isArray(svc.ports) ? svc.ports.map((p: any) => String(p)) : [];
    const environment = parseEnvironment(svc.environment);
    const replicasVal = svc.deploy?.replicas;
    const replicas = replicasVal !== undefined && replicasVal !== null ? parseInt(String(replicasVal), 10) : 1;
    const normalizedReplicas = Number.isNaN(replicas) ? 1 : replicas;
    const constraints: string[] = Array.isArray(svc.deploy?.placement?.constraints)
      ? svc.deploy.placement.constraints.map((c: any) => String(c))
      : [];

    result[name] = { image, ports, environment, replicas: normalizedReplicas, constraints };
  }

  return result;
}

function parseEnvironment(env: any): Record<string, string> {
  const result: Record<string, string> = {};

  if (Array.isArray(env)) {
    for (const item of env) {
      if (typeof item !== 'string') {
        console.warn(`Ignoring non-string environment entry: ${JSON.stringify(item)}`);
        continue;
      }
      const idx = item.indexOf('=');
      if (idx === -1) {
        console.warn(`Ignoring malformed environment entry: ${item}`);
        continue;
      }
      const key = item.slice(0, idx);
      const value = item.slice(idx + 1);
      result[key] = value;
    }
  } else if (typeof env === 'object' && env !== null) {
    for (const [key, value] of Object.entries(env)) {
      if (value === null) {
        result[key] = process.env[key] ?? '';
      } else {
        result[key] = String(value);
      }
    }
  }

  return result;
}

